From 378f23604f6e797e1bc951ec029f1857cc5d624b Mon Sep 17 00:00:00 2001
From: Sebastian Bauer <mail@sebastianbauer.info>
Date: Fri, 22 Jan 2016 20:04:50 +0100
Subject: [PATCH 08/42] Added libstc++ support for AmigaOS.

---
 .../os/{generic => amigaos}/ctype_base.h      |  0
 .../ctype_configure_char.cc                   |  4 +--
 .../os/{generic => amigaos}/ctype_inline.h    |  0
 .../os/{generic => amigaos}/error_constants.h | 28 +++++++++----------
 .../config/os/{djgpp => amigaos}/os_defines.h | 10 +++++--
 libstdc++-v3/configure.host                   |  3 ++
 6 files changed, 26 insertions(+), 19 deletions(-)
 copy libstdc++-v3/config/os/{generic => amigaos}/ctype_base.h (100%)
 copy libstdc++-v3/config/os/{bsd/darwin => amigaos}/ctype_configure_char.cc (98%)
 copy libstdc++-v3/config/os/{generic => amigaos}/ctype_inline.h (100%)
 copy libstdc++-v3/config/os/{generic => amigaos}/error_constants.h (92%)
 copy libstdc++-v3/config/os/{djgpp => amigaos}/os_defines.h (89%)

diff --git a/libstdc++-v3/config/os/generic/ctype_base.h b/libstdc++-v3/config/os/amigaos/ctype_base.h
similarity index 100%
copy from libstdc++-v3/config/os/generic/ctype_base.h
copy to libstdc++-v3/config/os/amigaos/ctype_base.h
diff --git a/libstdc++-v3/config/os/bsd/darwin/ctype_configure_char.cc b/libstdc++-v3/config/os/amigaos/ctype_configure_char.cc
similarity index 98%
copy from libstdc++-v3/config/os/bsd/darwin/ctype_configure_char.cc
copy to libstdc++-v3/config/os/amigaos/ctype_configure_char.cc
index b12c0bd435ca013496f95c0ed9dad9c884b20001..e8f5579caa949f20855a02cbde869d347d816e97 100644
--- a/libstdc++-v3/config/os/bsd/darwin/ctype_configure_char.cc
+++ b/libstdc++-v3/config/os/amigaos/ctype_configure_char.cc
@@ -39,14 +39,14 @@ _GLIBCXX_BEGIN_NAMESPACE_VERSION
 // Information as gleaned from /usr/include/ctype.h
 
   const ctype_base::mask*
   ctype<char>::classic_table() throw()
   { return 0; }
 
-  ctype<char>::ctype(__c_locale, const mask* __table, bool __del,
-		     size_t __refs)
+  ctype<char>::ctype(__c_locale, const mask* __table, bool __del, 
+          size_t __refs)
   : facet(__refs), _M_del(__table != 0 && __del),
   _M_toupper(NULL), _M_tolower(NULL),
   _M_table(__table ? __table : classic_table())
   {
     memset(_M_widen, 0, sizeof(_M_widen));
     _M_widen_ok = 0;
diff --git a/libstdc++-v3/config/os/generic/ctype_inline.h b/libstdc++-v3/config/os/amigaos/ctype_inline.h
similarity index 100%
copy from libstdc++-v3/config/os/generic/ctype_inline.h
copy to libstdc++-v3/config/os/amigaos/ctype_inline.h
diff --git a/libstdc++-v3/config/os/generic/error_constants.h b/libstdc++-v3/config/os/amigaos/error_constants.h
similarity index 92%
copy from libstdc++-v3/config/os/generic/error_constants.h
copy to libstdc++-v3/config/os/amigaos/error_constants.h
index 1d37454ba2efa2579de109dc1c0d28887ed58ed2..700e71fa192e7f559e9cbdbcf5ce67519e7527a3 100644
--- a/libstdc++-v3/config/os/generic/error_constants.h
+++ b/libstdc++-v3/config/os/amigaos/error_constants.h
@@ -45,13 +45,13 @@ _GLIBCXX_BEGIN_NAMESPACE_VERSION
       already_connected = 			EISCONN,
       argument_list_too_long = 			E2BIG,
       argument_out_of_domain = 			EDOM,
       bad_address = 				EFAULT,
       bad_file_descriptor = 			EBADF,
 
-#ifdef EBADMSG
+#ifdef _GLIBCXX_HAVE_EBADMSG
       bad_message = 				EBADMSG,
 #endif
 
       broken_pipe = 				EPIPE,
       connection_aborted = 			ECONNABORTED,
       connection_already_in_progress = 		EALREADY,
@@ -65,13 +65,13 @@ _GLIBCXX_BEGIN_NAMESPACE_VERSION
       file_exists = 	       			EEXIST,
       file_too_large = 				EFBIG,
       filename_too_long = 			ENAMETOOLONG,
       function_not_supported = 			ENOSYS,
       host_unreachable = 			EHOSTUNREACH,
 
-#ifdef EIDRM
+#ifdef _GLIBCXX_HAVE_EIDRM
       identifier_removed = 			EIDRM,
 #endif
 
       illegal_byte_sequence = 			EILSEQ,
       inappropriate_io_control_operation = 	ENOTTY,
       interrupted = 				EINTR,
@@ -83,92 +83,92 @@ _GLIBCXX_BEGIN_NAMESPACE_VERSION
       network_down = 				ENETDOWN,
       network_reset = 				ENETRESET,
       network_unreachable = 			ENETUNREACH,
       no_buffer_space = 			ENOBUFS,
       no_child_process = 			ECHILD,
 
-#ifdef ENOLINK
+#ifdef _GLIBCXX_HAVE_ENOLINK
       no_link = 				ENOLINK,
 #endif
 
       no_lock_available = 			ENOLCK,
 
-#ifdef ENODATA
+#ifdef _GLIBCXX_HAVE_ENODATA
       no_message_available = 			ENODATA,
 #endif
 
       no_message = 				ENOMSG,
       no_protocol_option = 			ENOPROTOOPT,
       no_space_on_device = 			ENOSPC,
 
-#ifdef ENOSR
+#ifdef _GLIBCXX_HAVE_ENOSR
       no_stream_resources = 			ENOSR,
 #endif
 
       no_such_device_or_address = 		ENXIO,
       no_such_device = 				ENODEV,
       no_such_file_or_directory = 		ENOENT,
       no_such_process = 			ESRCH,
       not_a_directory = 			ENOTDIR,
       not_a_socket = 				ENOTSOCK,
 
-#ifdef ENOSTR
+#ifdef _GLIBCXX_HAVE_ENOSTR
       not_a_stream = 				ENOSTR,
 #endif
 
       not_connected = 				ENOTCONN,
       not_enough_memory = 			ENOMEM,
 
-#ifdef ENOTSUP
+#ifdef _GLIBCXX_HAVE_ENOTSUP
       not_supported = 				ENOTSUP,
 #endif
 
-#ifdef ECANCELED
+#ifdef _GLIBCXX_HAVE_ECANCELED
       operation_canceled = 			ECANCELED,
 #endif
 
       operation_in_progress = 			EINPROGRESS,
       operation_not_permitted = 		EPERM,
       operation_not_supported = 		EOPNOTSUPP,
       operation_would_block = 			EWOULDBLOCK,
 
-#ifdef EOWNERDEAD
+#ifdef _GLIBCXX_HAVE_EOWNERDEAD
       owner_dead = 				EOWNERDEAD,
 #endif
 
       permission_denied = 			EACCES,
 
-#ifdef EPROTO
+#ifdef _GLIBCXX_HAVE_EPROTO
       protocol_error = 				EPROTO,
 #endif
 
       protocol_not_supported = 			EPROTONOSUPPORT,
       read_only_file_system = 			EROFS,
       resource_deadlock_would_occur = 		EDEADLK,
       resource_unavailable_try_again = 		EAGAIN,
       result_out_of_range = 			ERANGE,
 
-#ifdef ENOTRECOVERABLE
+#ifdef _GLIBCXX_HAVE_ENOTRECOVERABLE
       state_not_recoverable = 			ENOTRECOVERABLE,
 #endif
 
-#ifdef ETIME
+#ifdef _GLIBCXX_HAVE_ETIME
       stream_timeout = 				ETIME,
 #endif
 
-#ifdef ETXTBSY
+#ifdef _GLIBCXX_HAVE_ETXTBSY
       text_file_busy = 				ETXTBSY,
 #endif
 
       timed_out = 				ETIMEDOUT,
       too_many_files_open_in_system = 		ENFILE,
       too_many_files_open = 			EMFILE,
       too_many_links = 				EMLINK,
       too_many_symbolic_link_levels = 		ELOOP,
 
-#ifdef EOVERFLOW
+#ifdef _GLIBCXX_HAVE_EOVERFLOW
       value_too_large = 			EOVERFLOW,
 #elif defined __AVR__
       value_too_large = 			999,
 #endif
 
       wrong_protocol_type = 			EPROTOTYPE
diff --git a/libstdc++-v3/config/os/djgpp/os_defines.h b/libstdc++-v3/config/os/amigaos/os_defines.h
similarity index 89%
copy from libstdc++-v3/config/os/djgpp/os_defines.h
copy to libstdc++-v3/config/os/amigaos/os_defines.h
index 11c783871accd6e8f9bcbb076654e9c49e27b2cc..05d9dfa218aa597f89d9cad34d448aa4d3a250cd 100644
--- a/libstdc++-v3/config/os/djgpp/os_defines.h
+++ b/libstdc++-v3/config/os/amigaos/os_defines.h
@@ -1,7 +1,7 @@
-// Specific definitions for DJGPP  -*- C++ -*-
+// Specific definitions for AmigaOS  -*- C++ -*-
 
 // Copyright (C) 2001-2021 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
 // terms of the GNU General Public License as published by the
@@ -30,11 +30,15 @@
 #ifndef _GLIBCXX_OS_DEFINES
 #define _GLIBCXX_OS_DEFINES 1
 
 // System-specific #define, typedefs, corrections, etc, go here.  This
 // file will come before all others.
 
-// FIXME: should there be '#undef POSIX_SOURCE'?
+// No ioctl() on AmigaOS
+#define _GLIBCXX_NO_IOCTL 1
+ 
+#ifdef __NEWLIB__
+#define _GLIBCXX_USE_C99_STDINT_TR1 1
 
 #endif
 
-
+#endif
diff --git a/libstdc++-v3/configure.host b/libstdc++-v3/configure.host
index ec32980aa0db898623804980af65dad588e4d9f5..e9789a4981c6fc36c6b660ab768427bb42eb6509 100644
--- a/libstdc++-v3/configure.host
+++ b/libstdc++-v3/configure.host
@@ -213,12 +213,15 @@ case "${host_os}" in
     tmake_file="os/aix/t-aix"
     ;;
   aix*)
     os_include_dir="os/generic"
     atomicity_dir="cpu/generic"
     ;;
+  amigaos*)
+    os_include_dir="os/amigaos"
+    ;;
   bsd*)
     # Plain BSD attempts to share FreeBSD files.
     os_include_dir="os/bsd/freebsd"
     ;;
   cygwin*)
     os_include_dir="os/newlib"
-- 
2.34.1

